staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +34,445 @@ func (s *Source) ServiceNodeDelete(path string) {
+			return
+		default:
 		}
-		if provider {
-			otherChild, _, err := s.Con.Children(consumersPath)
-			if err != nil {
-				scope.Warnf("get children under the zk path %q failed: %s", consumersPath, err)
-			} else {
-				updateFunction(pathChildren, otherChild, path)
-			}
-		} else {
-			otherChild, _, err := s.Con.Children(providersPath)
-			if err != nil {
-				scope.Warnf("get children under the zk path %q failed: %s", consumersPath, err)
-			} else {
-				updateFunction(otherChild, pathChildren, path)
-			}
+
+		select {
+		case <-ctx.Done():
+			scope.Debugf("endpointWatcher %q exit due to ctx.Done()", ew.servicePath)
+			return
+		default:
 		}
-		for watchEvent := range pathEvent {
-			switch watchEvent.Type {
-			case zk.EventNodeDeleted:
-				if deleteFunction != nil {
-					deleteFunction(path)
-				}
-				s.watchPath.Remove(path)
-				scope.Infof("stop watch instance under the zk path %q because the node was deleted", path)
-				return
-			case zk.EventResyncWatch:
-				scope.Infof("instance under the zk path %q resync", path)
-				if provider {
-					if children, _, err := s.Con.Children(path); err == nil {
-						if consumers, _, err := s.Con.Children(consumersPath); err == nil {
-							updateFunction(children, consumers, providersPath)
-						}
-					}
-				} else {
-					if children, _, err := s.Con.Children(path); err == nil {
-						if providers, _, err := s.Con.Children(providersPath); err == nil {
-							updateFunction(providers, children, providersPath)
-						}
-					}
-				}
-			default:
-				// ignore other events as the channel will be closed unless the watch events are EventResyncWatch.
-				scope.Infof("instance under the zk path %q maybe changed, rewatch and update instance", path)
-			}
+
+		providers, _, err := ew.conn.Load().(*zk.Conn).Children(ew.servicePath + providerPathSuffix)
+		if err != nil {
+			scope.Warnf("endpointWatcher %q get current provider failed: %v", ew.servicePath, err)
+			ew.once.Do(ew.initCallback)
+			continue
 		}
+		// ignore errors when getting consumer fails in provider watcher
+		consumers, _, _ := ew.conn.Load().(*zk.Conn).Children(ew.servicePath + consumerPathSuffix)
+		ew.handler(providers, consumers)
+		ew.once.Do(ew.initCallback)
+		<-eventCh
 	}
 }
 
-var shouldStopWatchErrorMsg []string = []string{
-	"node does not exist",
+type ServiceEventQueue interface {
+	Add(ServiceEvent)
+	Len() int
+	Get() ServiceEvent
 }
 
-func shouldStopWatch(err error) bool {
-	for idx := range shouldStopWatchErrorMsg {
-		if strings.Contains(err.Error(), shouldStopWatchErrorMsg[idx]) {
-			return true
-		}
+func NewQueue() ServiceEventQueue {
+	return &eventQueue{
+		cond: sync.NewCond(&sync.Mutex{}),
 	}
-	return false
 }
 
-func (s *Source) EndpointUpdate(provider, consumer []string, path string) {
-	if !verifyPath(path, true) {
+type eventQueue struct {
+	cond *sync.Cond
+	q    []ServiceEvent
+}
+
+func (q *eventQueue) Add(item ServiceEvent) {
```

Comment:
enqueue

Reply:
the interface definition and implementation refer to k8s client-go, I will add description instead of renaming